packages/shared/src/react/contexts.tsx

```diff
@@ -1,28 +1,54 @@
 'use client';
 
 import type { PropsWithChildren } from 'react';
-import React from 'react';
+import React, { useState } from 'react';
 
 import type {
   BillingSubscriptionPlanPeriod,
   ClerkOptions,
-  ClientResource,
   ForPayerType,
+  InitialState,
   LoadedClerk,
   OrganizationResource,
-  SignedInSessionResource,
-  UserResource,
 } from '../types';
+import { useOrganizationBase } from './hooks/base/useOrganizationBase';
 import { createContextAndHook } from './hooks/createContextAndHook';
-import { SWRConfigCompat } from './providers/SWRConfigCompat';
 
 const [ClerkInstanceContext, useClerkInstanceContext] = createContextAndHook<LoadedClerk>('ClerkInstanceContext');
-const [UserContext, useUserContext] = createContextAndHook<UserResource | null | undefined>('UserContext');
-const [ClientContext, useClientContext] = createContextAndHook<ClientResource | null | undefined>('ClientContext');
-const [SessionContext, useSessionContext] = createContextAndHook<SignedInSessionResource | null | undefined>(
-  'SessionContext',
+export { useUserBase as useUserContext } from './hooks/base/useUserBase';
+export { useClientBase as useClientContext } from './hooks/base/useClientBase';
+export { useSessionBase as useSessionContext } from './hooks/base/useSessionBase';
+
+const [InitialStateContext, _useInitialStateContext] = createContextAndHook<InitialState | undefined>(
+  'InitialStateContext',
 );
 
+export function InitialStateProvider({
+  children,
+  initialState,
+}: {
+  children: React.ReactNode;
+  initialState: InitialState | undefined;
+}) {
+  // The initialState is not allowed to change, we snapshot it to turn that expectation into a guarantee.
+  // Note that despite this, it could still be different for different parts of the React tree which is fine,
+  // but that requires using a separate provider.
+  // eslint-disable-next-line react/hook-use-state
+  const [initialStateSnapshot] = useState(initialState);
+  const initialStateCtx = React.useMemo(() => ({ value: initialStateSnapshot }), [initialStateSnapshot]);
+  return <InitialStateContext.Provider value={initialStateCtx}>{children}</InitialStateContext.Provider>;
+}
+
+export function useInitialStateContext(): InitialState | undefined {
+  const initialState = _useInitialStateContext();
+
+  // If we want to support passing initialState as a promise, this is where we would handle that.
+  // Note that we can not use(promise) as long as we support React 18, so we'll need to have some extra handling
+  // and throw the promise instead.
```

Comment:
The idea is to implement this as a quick follow up. The reason we want this is to stop awaiting the `initialState` at the top level provider as one thing we should do to enable better PPR support.

I tried this with `use()` just to verify it worked, but an implementation with throwing is slightly trickier so wanted to do that as a separate PR.